modules/promotion/src/Element/CouponRedemptionForm.php

```diff
@@ -0,0 +1,225 @@
+    $wrapper_id = $id_prefix . '-ajax-wrapper';
+
+    $form_state->set('order', $element['#order']);
+    $element = [
+      '#tree' => TRUE,
+      '#prefix' => '<div id="' . $wrapper_id . '">',
+      '#suffix' => '</div>',
+      // Pass the id along to other methods.
+      '#wrapper_id' => $wrapper_id,
+    ] + $element;
+    $element['code'] = [
+      '#type' => 'textfield',
+      '#title' => t('Coupon code'),
+      '#description' => t('Enter your coupon code here.'),
+    ];
+    $element['apply'] = [
+      '#type' => 'submit',
+      '#value' => t('Apply coupon'),
+      '#name' => 'apply_coupon',
+      '#limit_validation_errors' => [
+        array_merge($element['#parents'], ['code']),
+      ],
+    ];
+
+    return $element;
+  }
+
+  /**
+   * Validates the coupon redemption form.
+   *
+   * @param array $element
+   *   The form element.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   The current state of the form.
+   */
+  public static function validateForm(array &$element, FormStateInterface $form_state) {
+    $coupon_parents = array_merge($element['#parents'], ['code']);
+    $coupon_code = $form_state->getValue($coupon_parents);
+    if (empty($coupon_code)) {
+      return;
+    }
+    $code_path = implode('][', $coupon_parents);
+    /** @var \Drupal\commerce_order\Entity\OrderInterface $order */
+    $order = $form_state->get('order');
+    $entity_type_manager = \Drupal::entityTypeManager();
+
+    // Validate that the coupon code exists.
+    $coupon = static::loadCouponByCode($coupon_code);
+    if (empty($coupon)) {
+      $form_state->setErrorByName($code_path, t('Coupon is invalid'));
+      return;
+    }
+
+    foreach ($order->get('coupons') as $item) {
+      if ($item->target_id == $coupon->id()) {
+        $form_state->setErrorByName($code_path, t('Coupon is already redeemed'));
+        return;
+      }
+    }
+
+    // If it does, ensure its promotion is still valid.
+    $order_type_storage = $entity_type_manager->getStorage('commerce_order_type');
+    /** @var \Drupal\commerce_promotion\PromotionStorageInterface $promotion_storage */
+    $promotion_storage = $entity_type_manager->getStorage('commerce_promotion');
+
+    /** @var \Drupal\commerce_order\Entity\OrderTypeInterface $order_type */
+    $order_type = $order_type_storage->load($order->bundle());
+    $promotion = $promotion_storage->loadByCoupon($order_type, $order->getStore(), $coupon);
+    if (empty($promotion)) {
+      $form_state->setErrorByName($code_path, t('Coupon is invalid'));
+      return;
+    }
+
+    $promotion_applies = FALSE;
+    if ($promotion->applies($order)) {
+      $promotion_applies = TRUE;
+    }
+    else {
+      foreach ($order->getItems() as $orderItem) {
+        if ($promotion->applies($orderItem)) {
+          $promotion_applies = TRUE;
+          break;
+        }
+      }
+    }
+
+    if ($promotion_applies === FALSE) {
```

Comment:
Why the strict comparison, when you're the one setting the boolean above? You should be able to trust it, no?